## Exercise - Option Price Introduction

The purpose of this exercise is to introduce you to a small amount of options data.  You'll perform a few calculations (which aren't all that meaningful) but the real motivation is to get your hands  dirty with some real-world data.

Along the way you will use some *indexing* and *querying* skills.

We will use the data in *spy_put.csv*

#### 1) Import the `pandas` and `numpy` packages.

In [1]:
import pandas as pd
import numpy as np

#### 2) Read-in the data into a variable call `df_put`.

In [2]:
df_put = pd.read_csv('spy_put.csv')
df_put.head()

underlying_symbol  underlying_price type  expiration   data_date  d2x  \
0               SPY        276.179688  put  2018-11-16  2018-10-19   20   
1               SPY        275.029999  put  2018-11-16  2018-10-22   19   
2               SPY        273.519989  put  2018-11-16  2018-10-23   18   
3               SPY        265.209991  put  2018-11-16  2018-10-24   17   
4               SPY        270.010010  put  2018-11-16  2018-10-25   16   

   strike   bid    ask    mid  volume  open_interest  implied_vol     delta  
0     274  4.24   4.27  4.255    7027          58931     0.174777  0.417679  
1     274  4.28   4.31  4.295    3174          58861     0.167284  0.441894  
2     274  5.09   5.14  5.115   32932          59467     0.176492  0.488031  
3     274  9.59  10.20  9.895   11461          49153     0.190524  0.703797  
4     274  8.11   8.19  8.150    3394          44592     0.196509  0.632242

For those of you unfamiliar with options, this might seem a little overwhelming, but don't worry.  You will be able to complete the exercises without any prior knowledge of options.

Options are insurance contracts written on top of shares of stocks.  Unlike stocks, options have an expiration date - on that date they either have some positive payout, or a they are worth nothing.  After expiration they cease to exist.  In `df_put` we are looking at end-of-day prices, for one particular option, for the 21 days before it expires.

In particular, this is a **put** option that has SPY as its **underlying**. The **expiration date** of our option is 11/16/2018.  The **strike price** of the contract is 274.  

Don't worry if this all sounds like gibberish.

#### 3) Use fancy indexing to create a `DataFrame` that consists of the put's contract features: `underlying_symbol`, `type`, `expiration`, and `strike`.  This won't be a very interesting `DataFrame` because all 21 rows will be the same.

In [3]:
df_put[['underlying_symbol', 'type', 'expiration', 'strike']].head() # alternative 1

underlying_symbol type  expiration  strike
0               SPY  put  2018-11-16     274
1               SPY  put  2018-11-16     274
2               SPY  put  2018-11-16     274
3               SPY  put  2018-11-16     274
4               SPY  put  2018-11-16     274

In [4]:
df_put.loc[:, ['underlying_symbol', 'type', 'expiration', 'strike']].head() # alternative 2

underlying_symbol type  expiration  strike
0               SPY  put  2018-11-16     274
1               SPY  put  2018-11-16     274
2               SPY  put  2018-11-16     274
3               SPY  put  2018-11-16     274
4               SPY  put  2018-11-16     274

#### 4) Create a `Series` of booleans, call it `ser_itm`,  by performing a pair-wise comparison of  `df_put['underlying_price']` and `df_put['strike']`.  In particular, check when the underlying price is less than the strike price of the option. 

In [5]:
ser_itm = df_put['underlying_price'] < df_put['strike']
ser_itm.head()

0    False
1    False
2     True
3     True
4     True
dtype: bool

#### 5) A put is said to be "in-the-money" if the price of the `underlying_price` is less than `strike`.  Use masking and `.query()` to select only the rows from `df_put` when the contract is in-the-money.

In [6]:
df_put[ser_itm].head()

underlying_symbol  underlying_price type  expiration   data_date  d2x  \
2               SPY        273.519989  put  2018-11-16  2018-10-23   18   
3               SPY        265.209991  put  2018-11-16  2018-10-24   17   
4               SPY        270.010010  put  2018-11-16  2018-10-25   16   
5               SPY        265.220001  put  2018-11-16  2018-10-26   15   
6               SPY        263.690002  put  2018-11-16  2018-10-29   14   

   strike    bid    ask     mid  volume  open_interest  implied_vol     delta  
2     274   5.09   5.14   5.115   32932          59467     0.176492  0.488031  
3     274   9.59  10.20   9.895   11461          49153     0.190524  0.703797  
4     274   8.11   8.19   8.150    3394          44592     0.196509  0.632242  
5     274   9.93  10.26  10.095    2387          44294     0.202849  0.711129  
6     274  11.39  11.86  11.625     741          43408     0.193238  0.792172

In [7]:
df_put.query('underlying_price < strike').head()

underlying_symbol  underlying_price type  expiration   data_date  d2x  \
2               SPY        273.519989  put  2018-11-16  2018-10-23   18   
3               SPY        265.209991  put  2018-11-16  2018-10-24   17   
4               SPY        270.010010  put  2018-11-16  2018-10-25   16   
5               SPY        265.220001  put  2018-11-16  2018-10-26   15   
6               SPY        263.690002  put  2018-11-16  2018-10-29   14   

   strike    bid    ask     mid  volume  open_interest  implied_vol     delta  
2     274   5.09   5.14   5.115   32932          59467     0.176492  0.488031  
3     274   9.59  10.20   9.895   11461          49153     0.190524  0.703797  
4     274   8.11   8.19   8.150    3394          44592     0.196509  0.632242  
5     274   9.93  10.26  10.095    2387          44294     0.202849  0.711129  
6     274  11.39  11.86  11.625     741          43408     0.193238  0.792172

#### 6) Calculate the average `delta` for days that the put is in-the-money, and also for the days that it is out-of-the-money.  (Hint: find out how to "negate" an array of booleans, apply your masking skills, and then apply an aggregation function).

In [8]:
df_put[ser_itm]['delta'].mean()

0.653652024181308

In [9]:
df_put[~(ser_itm)]['delta'].mean()

0.31517216213921667

In [10]:
df_put.query('underlying_price < strike')['delta'].mean()

0.653652024181308

In [11]:
df_put.query('underlying_price >= strike')['delta'].mean()

0.31517216213921667